### ISSUE TITLE
[BUG] Loss Overflow For GPT Neo Models When Adding New Tokens For All Precisions

### ISSUE THREAD

**Issue by mallorbc** [microsoft/DeepSpeed#2273]
**Describe the bug**

I am working on a task where I finetune GPT models.  I have tried GPT Neo and GPTJ models.  If I try to add custom token to GPT Neo models, I get an overflow error, and the loss scale tries to adjust(regardless if I am using fp16 or not).  I am able to train GPT J with additional tokens, I suspect due to already being 50400 for its output size.

Overflow for GPT Neo models occurs regardless of whether I am using fp16, fp32 or bf16. I have tried both Zero 2 and Zero 3 optimization, both of which work as expected when no new tokens are added.  Reading about this online, it seems to me like this should be possible.

**To Reproduce**
1. Create an instance of a GPT Neo model
2. Create an instance of the GPT Neo tokenizer
3. Add more tokens to the tokenizer(I added 6 tokens)
4. Train the model in any of the supported precision modes and watch observe loss overflows

**Expected behavior**

Limited to no overflow for fp16(until scaler is adjusted), no overflow for other formats.  

**ds_report output**
ds_report 
--------------------------------------------------
DeepSpeed C++/CUDA extension op report
--------------------------------------------------
NOTE: Ops not installed will be just-in-time (JIT) compiled at
      runtime if needed. Op compatibility means that your system
      meet the required dependencies to JIT install the op.
--------------------------------------------------
JIT compiled ops requires ninja
ninja .................. [OKAY]
--------------------------------------------------
op name ................ installed .. compatible
--------------------------------------------------
cpu_adam ............... [YES] ...... [OKAY]
cpu_adagrad ............ [YES] ...... [OKAY]
fused_adam ............. [YES] ...... [OKAY]
fused_lamb ............. [YES] ...... [OKAY]
sparse_attn ............ [YES] ...... [OKAY]
transformer ............ [YES] ...... [OKAY]
stochastic_transformer . [YES] ...... [OKAY]
async_io ............... [YES] ...... [OKAY]
utils .................. [YES] ...... [OKAY]
quantizer .............. [YES] ...... [OKAY]
transformer_inference .. [YES] ...... [OKAY]
--------------------------------------------------
DeepSpeed general environment info:
torch install path ............... ['/root/anaconda3/envs/gpt/lib/python3.9/site-packages/torch']
torch version .................... 1.12.0
torch cuda version ............... 11.3
torch hip version ................ None
nvcc version ..................... 11.3
deepspeed install path ........... ['/root/anaconda3/envs/gpt/lib/python3.9/site-packages/deepspeed']
deepspeed info ................... 0.7.1+7d8ad45, 7d8ad45, master
deepspeed wheel compiled w. ...... torch 1.12, cuda 11.3

**System info (please complete the following information):**
 - OS: Ubuntu 20.04
 - GPU count and types: 2 3090s
 - Interconnects: 1 system, 2 3090s
 - Python version: 3.9.13

I am using a docker container with Nvidia Cuda already set up as the base image.

**Launcher context**
fp16:
deepspeed --num_gpus=2 run_clm.py --deepspeed ds_config_gptneo.json --model_name_or_path EleutherAI/gpt-neo-2.7B --train_file train.csv --validation_file validation.csv --do_train --do_eval --overwrite_cache --evaluation_strategy=steps --output_dir finetuned --num_train_epochs 12  --eval_steps 10 --gradient_accumulation_steps 16 --per_device_train_batch_size 2 --use_fast_tokenizer False --learning_rate 5e-06 --warmup_steps 10 --save_total_limit 1 --save_steps 10 --save_strategy steps --tokenizer_name gpt2 --load_best_model_at_end True --block_size=2048 --fp16

bf16:
deepspeed --num_gpus=2 run_clm.py --deepspeed ds_config_gptneo.json --model_name_or_path EleutherAI/gpt-neo-2.7B --train_file train.csv --validation_file validation.csv --do_train --do_eval --overwrite_cache --evaluation_strategy=steps --output_dir finetuned --num_train_epochs 12  --eval_steps 10 --gradient_accumulation_steps 16 --per_device_train_batch_size 2 --use_fast_tokenizer False --learning_rate 5e-06 --warmup_steps 10 --save_total_limit 1 --save_steps 10 --save_strategy steps --tokenizer_name gpt2 --load_best_model_at_end True --block_size=2048 --bf16

fp32:
deepspeed --num_gpus=2 run_clm.py --deepspeed ds_config_gptneo.json --model_name_or_path EleutherAI/gpt-neo-2.7B --train_file train.csv --validation_file validation.csv --do_train --do_eval --overwrite_cache --evaluation_strategy=steps --output_dir finetuned --num_train_epochs 12  --eval_steps 10 --gradient_accumulation_steps 16 --per_device_train_batch_size 2 --use_fast_tokenizer False --learning_rate 5e-06 --warmup_steps 10 --save_total_limit 1 --save_steps 10 --save_strategy steps --tokenizer_name gpt2 --load_best_model_at_end True --block_size=2048

**Docker context**

Are you using a specific docker image that you can share?
nvidia/cuda:11.3.1-devel-ubuntu20.04 
then I am building python packages into the container

**Additional context**
When using fp32 and bf 16 I get messages like:
[deepspeed] OVERFLOW! Rank 0 Skipping step. Attempted loss scale: 1.0, reducing to 1.0

I am using slightly modified  https://github.com/mallorbc/Finetune_GPTNEO_GPTJ6B which is based off https://github.com/Xirider/finetune-gpt2xl for finetuning

code:
```
tokens_to_add = ["token1","token2"]
 tokenizer.add_tokens(tokens_to_add)
 model.resize_token_embeddings(len(tokenizer))
```

**Comment by samadejacobs** [comment]
@mallorbc, fp32 does not have loss scale, so this should not be happening. We have verified on our side that there is no loss overflow. Could you please (1) rerun with latest deepspeed master and or (2) run without deepspeed or with deepspeed zero stage 0 and report back if this still a problem?

**Comment by mallorbc** [comment]
I will try again.  Thanks!

**Comment by samadejacobs** [comment]
Great! For `fp16`, you may also want to reduce `initial_scale_power` to smaller value, say 2.
